### ISSUE TITLE
max_client_disconnect setting causes allocation states to hang in a non-terminal state when a client disconnects/reconnects

### ISSUE THREAD

**Issue by louievandyke** [hashicorp/nomad#19729]
### Nomad version
Output from `nomad version`

see below for tests on multiple versions

### Operating system and Environment details

ubuntu linux 

### Issue

We are running a job in which allocation Desired is `run` and Status is `failed`. nomad system gc will not clean the allocations because they are not in a terminal state and do not transition to terminal.

### Reproduction steps

Create one nomad server and one client.

I have tested the functionality with an example job 

With 1.4.14, I try with restart attempts to 3 or 0, and after some time nomad allocation Desired becomes stop. 

```
job "holiday" {
  datacenters = ["dc1"]

  type = "service"

  constraint {
      distinct_hosts = true
  }

  update {
    max_parallel = 1
    health_check = "checks"
    min_healthy_time = "10s"
    healthy_deadline = "15m"
    progress_deadline = "1h"
    auto_revert = false
    canary = 0
  }

  group "holiday_group" {
    count = 1
    max_client_disconnect = "1m"

    restart {
      attempts = 3
      interval = "1m"
      delay = "15s"
      mode = "fail"
    }

    reschedule {
      attempts = 0
      interval = "1m"
      delay = "15s"
      delay_function = "fibonacci"
      max_delay = "360s"
      unlimited = "true"
    }

    network {
      port "db" {
        to = 6379
      }
    }


    task "holiday_maintenance_pretask" {
      driver = "docker"

      lifecycle {
        hook = "prestart"
        sidecar = false
      }

      resources {
        memory = 50
      }


      config {
        image          = "redis:7"
        ports          = ["db"]
        auth_soft_fail = true
      }

    }

  }

}
```

```
root@nc1:/etc/nomad.d/job# nomad job status holiday
ID            = holiday
Name          = holiday
Submit Date   = 2023-12-28T18:16:03+05:30
Type          = service
Priority      = 50
Datacenters   = dc1
Namespace     = default
Status        = pending
Periodic      = false
Parameterized = false

Summary
Task Group     Queued  Starting  Running  Failed  Complete  Lost  Unknown
holiday_group  0       0         0        4       0         0     0

Future Rescheduling Attempts
Task Group     Eval ID   Eval Time
holiday_group  0c26dcd2  16s from now

Latest Deployment
ID          = d37bc0c3
Status      = running
Description = Deployment is running

Deployed
Task Group     Desired  Placed  Healthy  Unhealthy  Progress Deadline
holiday_group  1        4       0        4          2023-12-28T19:16:03+05:30

Allocations
ID        Node ID   Task Group     Version  Desired  Status  Created    Modified
33a689fe  611b1a54  holiday_group  0        run      failed  33s ago    24s ago
83ea399d  611b1a54  holiday_group  0        stop     failed  1m7s ago   33s ago
a6843b3c  611b1a54  holiday_group  0        stop     failed  1m26s ago  1m7s ago
a6b7bc5c  611b1a54  holiday_group  0        stop     failed  1m45s ago  1m26s ago
root@nc1:/etc/nomad.d/job#
```

With 1.4.14 I used restart attempts to 3 and max_client_disconnect = "1m", stop the client agent, After 1 minutes I can see alloc Desired,Status become  (stop, lost) on server. If I start the client agent again then allocation is trying to run on the same client node.

```
root@ns1:/tmp# nomad job status holiday
ID            = holiday
Name          = holiday
Submit Date   = 2023-12-28T18:23:13+05:30
Type          = service
Priority      = 50
Datacenters   = dc1
Namespace     = default
Status        = pending
Periodic      = false
Parameterized = false

Summary
Task Group     Queued  Starting  Running  Failed  Complete  Lost  Unknown
holiday_group  10      0         0        0       0         1     0

Placement Failure
Task Group "holiday_group":
  * No nodes were eligible for evaluation
  * No nodes are available in datacenter "dc1"

Latest Deployment
ID          = b8f23763
Status      = running
Description = Deployment is running

Deployed
Task Group     Desired  Placed  Healthy  Unhealthy  Progress Deadline
holiday_group  10       1       0        0          2023-12-28T19:23:13+05:30

Allocations
ID        Node ID   Task Group     Version  Desired  Status  Created    Modified
2dd0ab28  611b1a54  holiday_group  0        stop     lost    3m54s ago  9s ago
root@ns1:/tmp# 
```

With 1.5.10 I used restart attempts to 3 and after some time nomad allocation Desired,Status is showing in run, pending state.  Even if tried with max_client_disconnect = "1m" , I can see only run/pending status. Nomad system gc is not able to clean failed allocation as it is not moving their state to terminal.

```
root@nc1:/etc/nomad.d/job# nomad job status holiday
ID            = holiday
Name          = holiday
Submit Date   = 2023-12-28T18:39:27+05:30
Type          = service
Priority      = 50
Datacenters   = dc1
Namespace     = default
Status        = running
Periodic      = false
Parameterized = false

Summary
Task Group     Queued  Starting  Running  Failed  Complete  Lost  Unknown
holiday_group  0       1         0        0       0         0     0

Latest Deployment
ID          = 5baab18b
Status      = running
Description = Deployment is running

Deployed
Task Group     Desired  Placed  Healthy  Unhealthy  Progress Deadline
holiday_group  1        1       0        0          2023-12-28T19:39:27+05:30

Allocations
ID        Node ID   Task Group     Version  Desired  Status   Created    Modified
67f2bd6c  a066558f  holiday_group  0        run      pending  2m59s ago  13s ago
root@nc1:/etc/nomad.d/job#
```

With 1.7.2 I used restart attempts to 3 and after some time nomad allocation Desired,Status is not changing to a stop state. Even if tried with max_client_disconnect = "1m" 

```
root@nc1:/etc/nomad.d/job# nomad job status holiday
ID            = holiday
Name          = holiday
Submit Date   = 2023-12-28T18:55:00+05:30
Type          = service
Priority      = 50
Datacenters   = dc1
Namespace     = default
Node Pool     = default
Status        = running
Periodic      = false
Parameterized = false

Summary
Task Group     Queued  Starting  Running  Failed  Complete  Lost  Unknown
holiday_group  0       1         0        0       0         0     0

Latest Deployment
ID          = eb869f65
Status      = running
Description = Deployment is running

Deployed
Task Group     Desired  Placed  Healthy  Unhealthy  Progress Deadline
holiday_group  1        1       0        0          2023-12-28T19:55:00+05:30

Allocations
ID        Node ID   Task Group     Version  Desired  Status   Created   Modified
e2999ca8  57e214fd  holiday_group  0        run      pending  3m3s ago  14s ago
root@nc1:/etc/nomad.d/job#
```

#### Expected Result

After some time nomad allocation Desired becomes `stop`

#### Actual Result

The status for allocations does not change even when the count is fulfilled.

### Job file (if appropriate)

see above

**Comment by tgross** [comment]
@louievandyke you've provided a lot of extraneous detail here, like garbage collection behavior (which seems correct to me given the allocation states). But it looks like the core issue you're reporting is simply that `max_client_disconnect` field isn't being correctly respected from 1.5.10 onwards, inasmuch as an allocation with `reschedule.attempts = 0` should be marked `lost` once the `max_client_disconnect` window expires.

I know that @Juanadelacuesta fixed a bug in this behavior in https://github.com/hashicorp/nomad/pull/18701 (which appears to be missing a changelog entry). It would be helpful to know if we're seeing the evaluations we expect.

**Comment by vaiaro** [comment]
@tgross: Garbage collection behavior (which seems correct to me given the allocation states).Issue is while running the command "nomad job status example",Allocation desired,status were showing run,failed and desired,status is not moving to stop, complete. 

Can you please tell us what should we do to mark the allocation desired into stop state. nomad system gc will not clean the allocations because they are not in a terminal state and do not transition to terminal.

**Comment by tgross** [comment]
@vaiaro the bug is that this isn't happening automatically as we'd expect. As a workaround, you could `nomad alloc stop` those allocations, so long as the node the allocation is on is connected.

**Comment by Juanadelacuesta** [comment]
Fix mered and backported to 1.7, 1.6 and 1.5
